hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheRecordStore.java

```diff
@@ -164,64 +162,72 @@ public AbstractCacheRecordStore(String cacheNameWithPrefix, int partitionId, Nod
-            registerResourceIfItIsClosable(cacheLoader);
-            registerResourceIfItIsClosable(defaultExpiryPolicy);
-            init();
-        } finally {
-            try {
-                tenantContext.close();
-            } catch (IOException ex) {
-                ExceptionUtil.rethrow(ex);
-            }
-        }
-    }
+        this.eventJournalConfig = cacheConfig.getEventJournalConfig();
+        this.evictionConfig = cacheConfig.getEvictionConfig();
+        if (evictionConfig == null) {
+            throw new IllegalStateException("Eviction config cannot be null!");
+        }
+        this.wanReplicationEnabled = cacheService.isWanReplicationEnabled(cacheNameWithPrefix);
+        this.disablePerEntryInvalidationEvents = cacheConfig.isDisablePerEntryInvalidationEvents();
+
+        EvictionPolicyComparator evictionPolicyComparator = createEvictionPolicyComparator(evictionConfig);
+        evictionPolicyComparator = injectDependencies(evictionPolicyComparator);
+        this.evictionPolicyEvaluator = new EvictionPolicyEvaluator<>(evictionPolicyComparator);
+        this.cacheContext = cacheService.getOrCreateCacheContext(cacheNameWithPrefix);
+        this.records = createRecordCacheMap();
+        this.evictionChecker = createCacheEvictionChecker(evictionConfig.getSize(), evictionConfig.getMaxSizePolicy());
+        this.evictionStrategy = createEvictionStrategy(evictionConfig);
+        this.objectNamespace = CacheService.getObjectNamespace(cacheNameWithPrefix);
+        this.persistWanReplicatedData = canPersistWanReplicatedData(cacheConfig, nodeEngine);
+        this.cacheRecordFactory = new CacheRecordFactory(cacheConfig.getInMemoryFormat(), ss);
+        this.valueComparator = getValueComparatorOf(cacheConfig.getInMemoryFormat());
+        this.clearExpiredRecordsTask = cacheService.getExpirationManager().getTask();
 
-    private void initializeStatisticsAndFactories(String cacheNameWithPrefix) {
         if (cacheConfig.isStatisticsEnabled()) {
             statistics = cacheService.createCacheStatIfAbsent(cacheNameWithPrefix);
         }
-        if (cacheConfig.getCacheLoaderFactory() != null) {
-            Factory<CacheLoader> cacheLoaderFactory = cacheConfig.getCacheLoaderFactory();
-            cacheLoaderFactory = injectDependencies(cacheLoaderFactory);
-            cacheLoader = cacheLoaderFactory.create();
-            cacheLoader = injectDependencies(cacheLoader);
-        }
-        if (cacheConfig.getCacheWriterFactory() != null) {
-            Factory<CacheWriter> cacheWriterFactory = cacheConfig.getCacheWriterFactory();
-            cacheWriterFactory = injectDependencies(cacheWriterFactory);
-            cacheWriter = cacheWriterFactory.create();
-            cacheWriter = injectDependencies(cacheWriter);
-        }
+        injectDependencies(evictionPolicyEvaluator.getEvictionPolicyComparator());
+        TenantControl tenantControl = nodeEngine.getProxyService()
```

Comment:
Doesn't this undo the part of the effort of moving out tenant control out of the cache service? I thought the idea was to make it generic and not linked to cache.
I agree with you that it looks awkward however I don't know enough to see where it would be a better, generic place to put it that's accessible to all

Reply:
I basically need something 'global' where I can retrieve tenant control by service and distributed object names. Is there any generic place like that?

Reply:
It doesn't need to be just in the cache service, it can be part of any service that wishes to participate in tenant control. Some services may not need to participate in tenant control so hooking it into the proxy lifecycle might be unnecessary there (e.g. `FlakeIdGenerator`, `PNCounter`).

As for what would be the best "global" place, I guess it's just like any other helper subsystem (execution service, proxy service, ...) so to me `NodeEngine.getTenantControl` seems ok and generally `NodeEngine` is readily available almost everywhere.

Reply:
Ok, I can move it out of proxy into NodeEngine

Reply:
Let me correct that, I need to assess the feasibility of it first

Reply:
Ok, the problem I am having is that tenant control needs to go over the wire when distributed object is created.
This is why I put it into proxy anyway. Do you have the alternative of how to get tenant control created correctly with the distributed object while not using the proxy as the vehicle for this?

Reply:
Let me rephrase that. The 'specific' tenant control instance needs to be created when distributed object is created, and needs to go over the wire and be 'held' by the same distributed object everywhere in the cluster. This is why the code went into Proxy in the first place, because it is the 'mirror' of the distributed object on every node.

So the requirements are:
- tenant control holds state on a per-object basis, and needs to be distributed on the network
- tenant control needs to be kept with the object through all cluster nodes
- Objects, such as CacheRecordStore need object-specific tenantControl when created

Reply:
You're right, a proxy does seem like the perfect place for it then. Alternatively, it could be placed in a service, much like the proxy service, in `NodeEngine`, and share the same lifecycle, so that we could avoid mixing concerns in the proxy service and its' classes. Still, the proxy service and all of its' classes are private API, so we may simply ignore this complexity at this point, and merge it, only to consider moving it elsewhere in a subsequent PR.

Summary - I'll think about it a bit more but we can say this is not a blocker for the merge.